lib/Driver/DarwinToolChains.cpp

```diff
@@ -392,6 +393,35 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
   SmallString<128> RuntimeLibPath;
   getRuntimeLibraryPath(RuntimeLibPath, context.Args, /*Shared=*/true);
 
+  // Link the compatibility library, if we're deploying back to OSes that
+  // have a Swift 5.0 runtime.
+  
+  bool backDeployToSwift50;
+  unsigned Major, Minor, Micro;
+  
+  if (Triple.isMacOSX()) {
+    Triple.getMacOSXVersion(Major, Minor, Micro);
+    backDeployToSwift50 = Minor <= 14;
+  } else if (Triple.isiOS()) {
+    Triple.getiOSVersion(Major, Minor, Micro);
+    backDeployToSwift50 = Major <= 12;
+  } else if (Triple.isWatchOS()) {
+    Triple.getWatchOSVersion(Major, Minor, Micro);
+    backDeployToSwift50 = Major <= 5;
+  } else {
+    backDeployToSwift50 = false;
+  }
```

Comment:
Nitpick: can you factor this out to a helper function like the other deployment target checks?